rules.go

```diff
@@ -0,0 +1,482 @@
+	}
+}
+
+func parseDomainOption(text string) map[string]bool {
+	domains := text[len("domain="):]
+	parts := strings.Split(domains, "|")
+	opts := make(map[string]bool, len(parts))
+
+	for _, part := range parts {
+		opts[strings.TrimPrefix(part, "~")] = !strings.HasPrefix(part, "~")
+	}
+
+	return opts
+}
+
+func ruleToRegexp(text string) string {
+	// Convert AdBlock rule to a regular expression.
+	if text == "" {
+		return ".*"
+	}
+
+	// Check if the rule isn't already regexp
+	length := len(text)
+	if length >= 2 && text[:1] == "/" && text[length-1:] == "/" {
+		return text[1 : length-1]
+	}
+
+	// escape special regex characters
+	rule := text
+	rule = regexp.QuoteMeta(rule)
+
+	// |, ^ and * should not be escaped
+	rule = strings.ReplaceAll(rule, `\|`, `|`)
+	rule = strings.ReplaceAll(rule, `\^`, `^`)
+	rule = strings.ReplaceAll(rule, `\*`, `*`)
+
+	// XXX: the resulting regex must use non-capturing groups (?:
+	// for performance reasons; also, there is a limit on number
+	// of capturing groups, no using them would prevent building
+	// a single regex out of several rules.
+
+	// Separator character ^ matches anything but a letter, a digit, or
+	// one of the following: _ - . %. The end of the address is also
+	// accepted as separator.
+	rule = strings.ReplaceAll(rule, "^", `(?:[^\w\d_\\\-.%]|$)`)
+
+	// * symbol
+	rule = strings.ReplaceAll(rule, "*", ".*")
+
+	// | in the end means the end of the address
+	length = len(rule)
+	if rule[length-1] == '|' {
+		rule = rule[:length-1] + "$"
+	}
+
+	// || in the beginning means beginning of the domain name
+	if rule[:2] == "||" {
+		// XXX: it is better to use urlparse for such things,
+		// but urlparse doesn't give us a single regex.
+		// Regex is based on http://tools.ietf.org/html/rfc3986#appendix-B
+		if len(rule) > 2 {
+			//       |            | complete part       |
+			//       |  scheme    | of the domain       |
+			rule = `^(?:[^:/?#]+:)?(?://(?:[^/?#]*\.)?)?` + rule[2:]
+		}
+	} else if rule[0] == '|' {
+		// | in the beginning means start of the address
+		rule = "^" + rule[1:]
+	}
+
+	// other | symbols should be escaped
+	// we have "|$" in our regexp - do not touch it
+	rule = regexp.MustCompile(`(\|)[^$]`).ReplaceAllString(rule, `\|`)
+	return rule
+}
+
+func extractOptionsFromRequest(req Request) map[string]bool {
+	result := make(map[string]bool, len(supportedOptions))
+
+	filename := path.Base(req.URL.Path)
+	result["script"] = regexp.MustCompile(`(?:\.js$|\.js\.gz$)`).MatchString(filename)
+	result["image"] = regexp.MustCompile(`(?:\.jpg$|\.jpeg$|\.png$|\.gif$|\.webp$|\.tiff$|\.psd$|\.raw$|\.bmp$|\.heif$|\.indd$|\.jpeg2000$)`).MatchString(filename)
+	result["stylesheet"] = regexp.MustCompile(`(?:\.css$)`).MatchString(filename)
+	// More font extension at https://fileinfo.com/filetypes/font
+	result["font"] = regexp.MustCompile(`(?:\.otf|\.ttf|\.fnt)`).MatchString(filename)
+	result["thirdparty"] = req.Referer != ""
```

Comment:
So, I should considerer the Origin instead of Referer?
I used referer because looking at the behavior on network inspect from Firefox, shows a different referer if it comes from another website, and the same referer if is from same website